lib/rust/ensogl/core/src/display/shape/primitive/system.rs

```diff
@@ -534,6 +545,35 @@ where
 }
 
 
+// =================
+// === Parameter ===
+// =================
+
+/// A type which can be a shape parameter.
+///
+/// All types representable in Glsl (primitives, Vectors etc.) implements this by default.
+pub trait Parameter {
+    /// The type representation in GLSL. To be usable, it should implement [`Storable`] trait.
+    type GpuType;
+    /// The type representation in shader definition code.
+    type Variable;
+
+    /// A constructor of [`Self::Variable`] representing parameter with given name.
+    ///
+    /// The `name` should contain the obligatory `input_` prefix.
+    fn create_var(name: &str) -> Self::Variable;
+}
+
+impl<T: Storable> Parameter for T {
+    type GpuType = T;
+    type Variable = Var<T>;
+
+    default fn create_var(name: &str) -> Self::Variable {
+        name.into()
+    }
+}
+
```

Comment:
why do we need this? I mean, why we did not need it in the past and why we need it now? Based on the documentation I do not understand what is its purpose. I think extended docs will be helpful. Id love to help writing them but first I need to understand what is the "big picture" behind this abstraction

Reply:
See the implementation for `AnyCachedShape` - there we have different attribute than variable used in shape definition. It would not work in the old way.